### PR TITLE
KeyCloak Realm Service Clients

### PR DESCRIPTION
**Pull request by parc-jason**
Initial commit of KeyCloak Realm Administration

# Description
Add endpoint for creating/updating a Service Client in KeyCloak to use a Common Service (also in KeyCloak).
Update api documents for endpoint, and OpenShift documents for secrets and configuration.

## Types of changes

New feature (non-breaking change which adds functionality)
Documentation (non-breaking change with enhancements to documentation)

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-k8vopl-tools.pathfinder.gov.bc.ca/dashboard?id=nr-get-token-pr-82

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-k8vopl-tools.pathfinder.gov.bc.ca/dashboard?id=nr-get-token-pr-82

File: backend/src/docs/v1.api-spec.yaml [hunk 1]
```diff
@@ -299,6 +329,57 @@ components:
...
+        - applicationAcronym
+        - applicationName
+        - applicationDescription
+        - commonServices
+        - keycloakEnvironment
+        - passwordPublicKey
+      properties:
+        applicationAcronym:
+          type: string
+          example: MSSC
```
Comment (jujaga) [inline backend/src/docs/v1.api-spec.yaml]: Might be verbose, but we should always try to have a description field on every property even when it looks self evident.

File: openshift/backend.dc.yaml [hunk 2]
```diff
@@ -162,6 +162,16 @@ objects:
...
+          - name: SC_KC_INT_USERNAME
+            valueFrom:
+              secretKeyRef:
+                key: username
+                name: getok-sc-keycloakint-secret
+          - name: SC_KC_INT_PASSWORD
+            valueFrom:
+              secretKeyRef:
+                key: password
+                name: getok-sc-keycloakint-secret
```
Comment (jujaga) [inline openshift/backend.dc.yaml]: May need to update/rename based on environment variable name change

File: openshift/README.md [hunk 3]
```diff
@@ -31,7 +31,9 @@ oc create -n k8vopl-<env> configmap getok-sc-config \
   --from-literal=SC_GETOK_ENDPOINT_INT=https://i1api.nrs.gov.bc.ca/webade-api/v1 \
   --from-literal=SC_GETOK_ENDPOINT_TEST=https://t1api.nrs.gov.bc.ca/webade-api/v1 \
   --from-literal=SC_GETOK_ENDPOINT_PROD=https://api.nrs.gov.bc.ca/webade-api/v1 \
-  --from-literal=SC_MSSC_ENDPOINT=https://i1api.nrs.gov.bc.ca/cmsg-messaging-api/v1
+  --from-literal=SC_MSSC_ENDPOINT=https://i1api.nrs.gov.bc.ca/cmsg-messaging-api/v1 \
+  --from-literal=SC_KC_INT_ENDPOINT=https://sso-dev.pathfinder.gov.bc.ca \
+  --from-literal=SC_KC_INT_REALM=98r0z7rz
```
Comment (jujaga) [inline openshift/README.md]: May need to update/rename based on environment variable name change

File: openshift/README.md [hunk 4]
```diff
@@ -76,6 +78,14 @@ oc create -n k8vopl-<env> secret generic getok-sc-mssc-secret \
   --from-literal=password=<password>
 ```
 
+```sh
+oc create -n k8vopl-<env> secret generic getok-sc-keycloakint-secret \
+  --type=kubernetes.io/basic-auth \
+  --from-literal=username=<username> \
+  --from-literal=password=<password>
+```
```
Comment (jujaga) [inline openshift/README.md]: May need to update/rename based on environment variable name change

File: backend/src/routes/v1/kcClientForm.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+const {
+  body,
+  validationResult
+} = require('express-validator');
+
+// submits a keycloak service client
+// will need to add public key to this for encryption from frontend...
+const kcClientForm = require('express').Router();
+kcClientForm.post('/', [
```
Comment (jujaga) [inline backend/src/routes/v1/kcClientForm.js]: Missing whitespace line above

File: backend/src/routes/v1/kcClientForm.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    passwordPublicKey: publicKey
+  } = req.body;
+
+  try {
+    const realmKey = `serviceClient.keyCloak.${serviceClientForm.keycloakEnvironment}`;
+    const {
+      endpoint: realmBaseUrl,
+      username: clientId,
+      password: clientSecret,
+      realm: realmId} = config.get(realmKey);
```
Comment (jujaga) [inline backend/src/routes/v1/kcClientForm.js]: newline recommended between `realmId` and the `}`.

File: backend/config/custom-environment-variables.json [hunk 7]
```diff
@@ -37,6 +37,14 @@
...
       "username": "SC_MSSC_USERNAME",
       "password": "SC_MSSC_PASSWORD"
+    },
+    "keyCloak": {
+      "INT": {
+        "endpoint": "SC_KC_INT_ENDPOINT",
+        "username": "SC_KC_INT_USERNAME",
+        "password": "SC_KC_INT_PASSWORD",
+        "realm": "SC_KC_INT_REALM"
+      }
```
[the remark this answers is not in the document]
Comment (parc-jason) [reply backend/config/custom-environment-variables.json]: We discussed changing how to determine "environments", wanted to remove the "environment" from the name of the type. We will have different lists of target environments - one drop down list for WebADE, another for KeyCloak.  Nicer to have that hierarchy separation.

File: backend/src/components/realmAdminSvc.js [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+    this.axios.interceptors.request.use(
+      // Wraps axios-token-interceptor with oauth-specific configuration,
+      // fetches the token using the desired claim method, and caches
+      // until the token expires
+      oauth.interceptor(tokenProvider, oauth.client(axios.create(), {
+        url: this.tokenUrl,
+        grant_type: 'client_credentials',
+        client_id: clientId,
+        client_secret: clientSecret,
+        scope: ''
```
[the remark this answers is not in the document]
Comment (parc-jason) [reply backend/src/components/realmAdminSvc.js]: It is supposed to be empty in this case.  We don't have a scope on this service client.

File: backend/src/routes/v1/kcClientForm.js (hunk 5 above)
Comment (parc-jason) [reply backend/src/routes/v1/kcClientForm.js]: are these enforceable in es-lint?

File: backend/src/routes/v1/kcClientForm.js (hunk 6 above)
Comment (parc-jason) [reply backend/src/routes/v1/kcClientForm.js]: are these enforceable in es-lint?

File: backend/src/components/realmAdminSvc.js [hunk 9]
```diff
@@ -0,0 +1,372 @@
+const axios = require('axios');
+const oauth = require('axios-oauth-client');
+const tokenProvider = require('axios-token-interceptor');
+const log = require('npmlog');
```
[the remark this answers is not in the document]
Comment (parc-jason) [reply backend/src/components/realmAdminSvc.js]: ok - next commit

File: backend/src/routes/v1/kcClientForm.js (hunk 5 above)
Comment (parc-jason) [reply backend/src/routes/v1/kcClientForm.js]: addressed in next commit

File: backend/src/routes/v1/kcClientForm.js (hunk 6 above)
Comment (parc-jason) [reply backend/src/routes/v1/kcClientForm.js]: addressed in next commit

File: backend/src/components/realmAdminSvc.js [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+      }))
+    );
+  }
+
+  async getRealm() {
+    const response = await this.axios.get(this.realmAdminUrl)
+      .catch(e => {
+        log.error('RealmAdminService.getRealm', JSON.stringify(e));
+        throw e;
+      });
```
[the remark this answers is not in the document]
Comment (parc-jason) [reply backend/src/components/realmAdminSvc.js]: I only want to catch and log error from the promise here, nothing else.  although these class methods pretty much have no other business logic, and no other errors to throw, it's clear that the http error/exception is all that I want caught and logged.  The try/catch block is expected to be further up the call stack.

File: backend/src/components/realmAdminSvc.js [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+    return response.data;
+  }
+
+  async getClients() {
+    const response = await this.axios.get(`${this.realmAdminUrl}/clients`)
+      .catch(e => {
+        log.error('RealmAdminService.getClients', JSON.stringify(e));
+        throw e;
+      });
```
[the remark this answers is not in the document]
Comment (parc-jason) [reply backend/src/components/realmAdminSvc.js]: see above

File: backend/src/components/realmAdminSvc.js [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+  async getClient(id) {
+    if (!id) {
+      log.error('RealmAdminService getClient id parameter is null.');
+      throw new Error('Cannot get client: id parameter cannot be null.');
+    }
+    const response = await this.axios.get(`${this.realmAdminUrl}/clients/${id}`)
+      .catch(e => {
+        log.error('RealmAdminService.getClient', JSON.stringify(e));
+        throw e;
+      });
```
[the remark this answers is not in the document]
Comment (parc-jason) [reply backend/src/components/realmAdminSvc.js]: see above

File: backend/src/components/realmAdminSvc.js [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+      JSON.stringify(requestBody),
+      {
+        headers: {
+          'Content-Type': 'application/json'
+        }
+      }
+    ).catch(e => {
+      log.error('RealmAdminService.updateClientDetails', JSON.stringify(e));
+      throw e;
+    });
```
[the remark this answers is not in the document]
Comment (parc-jason) [reply backend/src/components/realmAdminSvc.js]: see above

File: backend/src/components/realmAdminSvc.js [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+    if (!id) {
+      log.error('RealmAdminService getClientSecret id parameter is null.');
+      throw new Error('Cannot get client secret: id parameter cannot be null.');
+    }
+
+    const response = await this.axios.get(`${this.realmAdminUrl}/clients/${id}/client-secret`
+    ).catch(e => {
+      log.error('RealmAdminService.getClientSecret', JSON.stringify(e));
+      throw e;
+    });
```
[the remark this answers is not in the document]
Comment (parc-jason) [reply backend/src/components/realmAdminSvc.js]: see above

File: backend/src/components/realmAdminSvc.js [hunk 15]
```diff
@@ -0,0 +1,372 @@
...
+    if (!id) {
+      log.error('RealmAdminService getServiceAccountUser id parameter is null.');
+      throw new Error('Cannot get client service account user: id parameter cannot be null.');
+    }
+
+    const response = await this.axios.get(`${this.realmAdminUrl}/clients/${id}/service-account-user`)
+      .catch(e => {
+        log.error('RealmAdminService.getServiceAccountUser', JSON.stringify(e));
+        throw e;
+      });
```
[the remark this answers is not in the document]
Comment (parc-jason) [reply backend/src/components/realmAdminSvc.js]: see above

File: backend/src/components/realmAdminSvc.js [hunk 16]
```diff
@@ -0,0 +1,372 @@
...
+    if (!id) {
+      log.error('RealmAdminService getClientRoles id parameter is null.');
+      throw new Error('Cannot get client roles: id parameter cannot be null.');
+    }
+
+    const response = await this.axios.get(`${this.realmAdminUrl}/clients/${id}/roles`
+    ).catch(e => {
+      log.error('RealmAdminService.getClientRoles', JSON.stringify(e));
+      throw e;
+    });
```
[the remark this answers is not in the document]
Comment (parc-jason) [reply backend/src/components/realmAdminSvc.js]: see above

File: backend/src/components/realmAdminSvc.js [hunk 17]
```diff
@@ -0,0 +1,372 @@
...
+      JSON.stringify(roles),
+      {
+        headers: {
+          'Content-Type': 'application/json'
+        }
+      }
+    ).catch(e => {
+      log.error('RealmAdminService.setRoleComposites', JSON.stringify(e));
+      throw e;
+    });
```
[the remark this answers is not in the document]
Comment (parc-jason) [reply backend/src/components/realmAdminSvc.js]: see above

File: backend/src/docs/v1.api-spec.yaml (hunk 1 above)
Comment (parc-jason) [reply backend/src/docs/v1.api-spec.yaml]: just following what was in the spec already. we know that the spec needs to be updated (endpoints that do not exist).  File this away for when the spec is updated to match production.

File: backend/src/routes/v1/kcClientForm.js (hunk 5 above)
Comment (jujaga) [reply backend/src/routes/v1/kcClientForm.js]: I don't think there's anything in es-lint regarding the above type of spacing unfortunately - at least I'm not aware of it anyways.

**Pull request merged**